adi/ad7606.py

```diff
@@ -0,0 +1,160 @@
+#       the documentation and/or other materials provided with the
+#       distribution.
+#     - Neither the name of Analog Devices, Inc. nor the names of its
+#       contributors may be used to endorse or promote products derived
+#       from this software without specific prior written permission.
+#     - The use of this software may or may not infringe the patent rights
+#       of one or more patent holders.  This license does not release you
+#       from the requirement that you obtain separate licenses from these
+#       patent holders to use this software.
+#     - Use of the software either in source or binary form, must be run
+#       on or directly connected to an Analog Devices Inc. component.
+#
+# THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+# INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT, MERCHANTABILITY AND FITNESS FOR A
+# PARTICULAR PURPOSE ARE DISCLAIMED.
+#
+# IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, INTELLECTUAL PROPERTY
+# RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+# BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+# THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+import numpy as np
+from adi.attribute import attribute
+from adi.context_manager import context_manager
+from adi.rx_tx import rx
+from decimal import Decimal
+
+
+class ad7606(rx, context_manager):
+    """ AD7606 ADC """
+
+    _complex_data = False
+    channel = []  # type: ignore
+    _device_name = ""
+
+    def __init__(self, uri="", device_name=""):
+
+        context_manager.__init__(self, uri, self._device_name)
+
+        compatible_parts = ["ad7605-4", "ad7606-4", "ad7606-6", "ad7606-8", "ad7606b", "ad7616"]
+
+        self._ctrl = None
+
+        if not device_name:
+            device_name = compatible_parts[0]
+        else:
+            if device_name not in compatible_parts:
+                raise Exception("Not a compatible device: " + device_name)
+
+        # Select the device matching device_name as working device
+        for device in self._ctx.devices:
+            if device.name == device_name:
+                self._ctrl = device
+                self._rxadc = device
+                break
+
+        for ch in self._ctrl._channels:
+            name = ch._id
+            self._rx_channel_names.append(name)
+            self.channel.append(self._channel(self._ctrl, name))
+
+        rx.__init__(self)
+
+    def rx(self):
+        sig = super().rx()
+
+        if (
+            self._rx_unbuffered_data
+            or self._complex_data
+            or self.rx_output_type == "raw"
+        ):
+            return sig
+        else:
```

Comment:
The else is unnecessary here. You would do this in a little list comprehension too like: 
```python
return [sig/1000 for sig in signal]
```

Reply:
I referred this code from ad7124.py module. Below is my justification for this part of code. Please correct me if I am wrong.
The 'if' part is used to return ADC read data just in raw format provided 'rx_output_type' type is "raw" Or '_rx_unbuffered_data' is true or '_complex_data' is true. I think '_complex_data' might be redundant here as it is local to ad7606 class and set to 'false' during init. May be this can be removed from 'if' check?
The other two attributes are defined in 'rx_tx.py'. So if someone want's read data in voltage converted form, they can set '_rx_unbuffered_data = true' Or 'rx_output_type = "SI"'. Again this is referred from ad7124.py. The else part would then convert raw data into eqv voltage (by considering scale factor).
Let me know your thoughts. I have added test results below (this is for my ARM based firmware example and not on FPGA):

![image](https://user-images.githubusercontent.com/62383520/109762009-4d200180-7c16-11eb-9a38-e810463c8923.png)

Reply:
Sorry my comment wasn't clear. The separate conditions are fine but since there is a return you don't need to use an else for the second condition. It should be something like:

if (self._rx_unbuffered_data
            or self._complex_data
            or self.rx_output_type == "raw"
        ):
     return sig

return [sig/1000 for sig in signal]

Reply:
Okay got it now, sorry for the confusion. I will make this change.
Scale attribute is not set to 1/10000. So I think that's why result needs to be divided by 1000.

https://wiki.analog.com/resources/tools-software/linux-drivers/iio-adc/ad7606
![image](https://user-images.githubusercontent.com/62383520/110061900-57fda200-7d8e-11eb-9070-4706bea7afaa.png)

Reply:
Sorry for the delay. I was out on vacation.

So after applying the scale and bias it should result in a millivolt. Is that not the case?

Reply:
By looking at AD7606 Linux drivers, the scale is represented to obtain the results in millivolt. 
For example: scale = range / adc resolution = 10v / 2^16 = 10/65535 = 0.000152
However, range factor is multiplied by 1000 in Linux drivers, so the scale becomes = 10000/65535 = 0.152.
And then formula used is: U = in0_raw * in_scale = 5789 * 0.152 = 879.928 mV

So, your comment above is valid i.e. after multiplying scale with adc raw value, the result should be in millivolts. And if that is the case, the additional code used to convert raw value into millivolt (i.e. signal/1000) is unnecessary.

The function would then look like as below, right?

**def rx(self):
        return super().rx()**

I would wait for your inputs on this before committing code, just to make sure my understanding is correct.

Reply:
kk. So if that is the case you don't even need the rx method in your class since it will just inherit it.

-Travis

Reply:
Agree with you, this method won't require. Do you want me to delete this method completely from ad7606.py class file?

Reply:
Yes. That should be the last thing and I'll merge. Thanks for the contribution :+1: 

-Travis